cmd/runtimetest/main.go

```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
 	}
 
 	for fs, fstype := range defaultFS {
-		if !(mountsMap[fs] == fstype) {
-			return specerror.NewError(specerror.DefaultFilesystems, fmt.Errorf("%v SHOULD exist and expected type is %v", fs, fstype), rspec.Version)
+		rfcError, err := c.Ok(mountsMap[fs] == fstype, specerror.DefaultFilesystems, spec.Version, fmt.Sprintf("mount %v has expected type", fs))
+		if err != nil {
+			return err
 		}
+		c.harness.YAML(map[string]string{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"mount":     fs,
+			"expected":  fstype,
+			"actual":    mountsMap[fs],
+		})
 	}
 
 	return nil
 }
 
-func validateLinuxDevices(spec *rspec.Spec, t *tap.T) error {
-	if spec.Linux == nil {
+func (c *complianceTester) validateLinuxDevices(spec *rspec.Spec) error {
+	if spec.Linux == nil || spec.Linux.Devices == nil {
+		c.harness.Skip(1, "linux.devices is unset")
 		return nil
 	}
-	for _, device := range spec.Linux.Devices {
-		fi, err := os.Stat(device.Path)
+
+	for i, device := range spec.Linux.Devices {
+		err := c.validateDevice(
+			&device,
+			specerror.DevicesAvailable,
+			spec.Version,
+			fmt.Sprintf("%q (linux.devices[%d])", device.Path, i))
 		if err != nil {
 			return err
 		}
-		fStat, ok := fi.Sys().(*syscall.Stat_t)
-		if !ok {
-			return specerror.NewError(specerror.DevicesAvailable, fmt.Errorf("cannot determine state for device %s", device.Path), rspec.Version)
-		}
-		var devType string
-		switch fStat.Mode & syscall.S_IFMT {
-		case syscall.S_IFCHR:
-			devType = "c"
-		case syscall.S_IFBLK:
-			devType = "b"
-		case syscall.S_IFIFO:
-			devType = "p"
-		default:
-			devType = "unmatched"
-		}
-		if devType != device.Type || (devType == "c" && device.Type == "u") {
-			return fmt.Errorf("device %v expected type is %v, actual is %v", device.Path, device.Type, devType)
-		}
-		if devType != "p" {
-			dev := fStat.Rdev
-			major := (dev >> 8) & 0xfff
-			minor := (dev & 0xff) | ((dev >> 12) & 0xfff00)
-			if int64(major) != device.Major || int64(minor) != device.Minor {
-				return fmt.Errorf("%v device number expected is %v:%v, actual is %v:%v", device.Path, device.Major, device.Minor, major, minor)
-			}
+	}
+
+	return nil
+}
+
+func (c *complianceTester) validateDevice(device *rspec.LinuxDevice, condition specerror.Code, version string, description string) (err error) {
+	var exists bool
+	fi, err := os.Stat(device.Path)
+	if os.IsNotExist(err) {
+		exists = false
+	} else if err != nil {
+		return err
+	} else {
+		exists = true
+	}
+	rfcError, err := c.Ok(exists, specerror.DevicesAvailable, version, fmt.Sprintf("has a file at %s", description))
```

Comment:
`specerror.DevicesAvailable` -> `condition`

Reply:
ditto in the 6 other occurrences of `specerror.DevicesAvailable` below in this function

Reply:
> `specerror.DevicesAvailable` -> `condition`

Thanks.  I've fixed all of those with 06482da → 031e4c8.